app/src/main/java/net/bonysoft/magicmirror/facerecognition/KeyboardFaceSource.java

```diff
@@ -0,0 +1,57 @@
+package net.bonysoft.magicmirror.facerecognition;
+
+public class KeyboardFaceSource implements FaceReactionSource {
+
+    private static final FaceExpression NEUTRAL_EXPRESSION = FaceExpression.LOOKING;
+    private static final int KEY_NEUTRAL = -1;
+
+    private final FaceTracker.FaceListener faceListener;
+    private final KeyToFaceMappings mappings;
+
+    private int currentPress = KEY_NEUTRAL;
+
+    public KeyboardFaceSource(FaceTracker.FaceListener faceListener, KeyToFaceMappings mappings) {
+        this.faceListener = faceListener;
+        this.mappings = mappings;
+    }
+
+    public boolean onKeyDown(int keyCode) {
+        if (sameKeyCodeIsTriggered(keyCode) || isStillUnsupportedCode(keyCode)) {
+            return false;
+        }
+        FaceExpression faceExpression = mappings.getFace(keyCode);
+        if (faceExpression != null) {
+            currentPress = keyCode;
+            faceListener.onNewFace(faceExpression);
+            return true;
+        }
+        currentPress = KEY_NEUTRAL;
+        return false;
+    }
+
+    private boolean sameKeyCodeIsTriggered(int keyCode) {
+        return currentPress == keyCode;
+    }
+
+    private boolean isStillUnsupportedCode(int keyCode) {
+        return currentPress == KEY_NEUTRAL && !mappings.contains(keyCode);
+    }
+
+    public boolean onKeyUp(int keyCode) {
+        if (sameKeyCodeIsTriggered(keyCode)) {
+            faceListener.onNewFace(NEUTRAL_EXPRESSION);
+            currentPress = KEY_NEUTRAL;
+        }
+        return true;
+    }
+
+    @Override
+    public void start() {
+        // no-op
```

Comment:
That is true. We had a chat with @danybony on how do this properly and came up with the idea to create two separate activities with each separate source (as Fragments do not have a `onKeyUp/Down()`). If you have one activity, you end up having code on `onResume/Stop` for the Camera preview and on `onKeyDown/Up`.

As the available time we have is really limited and _nothing_ will be built on top of this, we decided to go like this.